### PR TITLE
Support for passReqToCallback option

### PR DESCRIPTION
**Pull request by S-Stephen**
Dear Forbes,

(I'm sure the if statement could be written a little more succinctly. )

Whilst attempting to try out your passport-raven module in a Sails application I needed to access the request during the callback.  I noticed that there was an option I could set (passReqToCallback) but your modules didn't currently support it.  I added in here so hopefully it can now provide this feature.

The change to the test script sadly does not include a test for this change bu the ability to test the system from another host rather than localhost (I've no GUI on my dev machine).

I'd be grateful if you could consider these changes.  I'm, hopefully, intending to use them

Thanks in advance,
Stephen

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -45,6 +45,25 @@ passport.use(new RavenStrategy({
  - `desc` (String) - a description of the website to be displayed on the raven login page
  - `msg` (String) - a description of why authentication is being requested
  - `iact` (Boolean) - Set to `true` to force users to type their username and password even if they are already logged in. Set to `false` to only login if it can be done without user interaction.  Defaults to `null`.
+ - `passReqToCallback` (Boolean) - If set the request object is provided as the first argument to the verify function. The verify callback can therefor euse the state of the request to tailer further handling.
```
Comment (ForbesLindesay) [inline README.md]: "therefore use"

File: README.md [hunk 2]
```diff
@@ -45,6 +45,25 @@ passport.use(new RavenStrategy({
...
+
+```js
+
+passport.use(new RavenStrategy({
+  desc: 'My Raven Application',
+  msg: 'we need to check you are a current student',
+  // use demonstration raven server in development
+  debug: process.env.NODE_ENV !== 'production',
+  passReqToCallback : true
+}, function (req,crsid, params, callback) {
```
Comment (ForbesLindesay) [inline README.md]: Missing space after `,`

File: index.js [hunk 3]
```diff
@@ -107,11 +107,19 @@ Strategy.prototype.processResponse = function (req) {
       if (checkSignature(data, response.sig, this.debug ? KEYS.debug : KEYS.production)) {
         debug('Raven response signature check passed.');
         response.isCurrent = response.ptags === 'current';
-        return self._verify(response.principal, response, function (err, user, info) {
-          if (err) { return self.error(err); }
-          if (!user) { return self.fail(info); }
-          self.success(user, info);
-        });
+        if (self._opts['passReqToCallback']) {
+        	return self._verify(req,response.principal, response, function (err, user, info) {
```
Comment (ForbesLindesay) [inline index.js]: There should be a space after `req,` and the indentation looks way off here.  We're using 2 spaces for indents.

File: test/server.js [hunk 4]
```diff
@@ -5,7 +5,7 @@ var passport = require('passport');
 var Raven = require('../');
 
 passport.use(new Raven({
-  audience: 'http://localhost:3000',
+  audience: process.env.test_audience || 'http://localhost:3000',
```
Comment (ForbesLindesay) [inline test/server.js]: Why are you adding an environment variable for this?

**Comment by ForbesLindesay** [comment]
What's the purpose of this?  What information is in the request that you need access to within the verify callback?  Is this a common pattern that is in use for other passport strategies?  If so, can you point me at some examples?

**Comment by S-Stephen** [comment]
Sure example usage is perhaps here:

http://passportjs.org/docs

Further to making sure that the user is a member of the university I also hold application level authorisation levels in the application.  The frame work I'm using is the Sails with a passport service provider using a connect function (passport.connect) that requires the request object:

https://github.com/kasperisager/sails-generate-auth/blob/master/templates/api/services/passport.js

see comment:
- As you can see, this function handles both "authentication" and "authori-
  - zation" at the same time. This is due to the fact that we pass in
  - `passReqToCallback: true` when loading the strategies, allowing us to look
  - for an existing session in the request and taking action based on that.

**Comment by ForbesLindesay** [comment]
OK, I'm convinced :+1: Fix up those nits and I'll merge it in.

**Comment by S-Stephen** [comment]
Hopefully all resolved, though not sure it helps you can no longer see the comments for the diffs?!

I removed the env variable in test/server.js

**Pull request merged**